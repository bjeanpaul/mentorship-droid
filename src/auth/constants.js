const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
const AUTH_LOGIN_BUSY = 'AUTH_LOGIN_BUSY';
const AUTH_LOGIN_SUCCESS = 'AUTH_LOGIN_SUCCESS';
const AUTH_LOGIN_FAILURE = 'AUTH_LOGIN_FAILURE';
const AUTH_PASSWORD_RESET_BUSY = 'AUTH_PASSWORD_RESET_BUSY';
const AUTH_PASSWORD_RESET_SUCCESS = 'AUTH_PASSWORD_RESET_SUCCESS';
const AUTH_PASSWORD_RESET_FAILURE = 'AUTH_PASSWORD_RESET_FAILURE';

export {
  AUTH_SET_TOKEN,
  AUTH_LOGIN_BUSY,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_PASSWORD_RESET_BUSY,
  AUTH_PASSWORD_RESET_SUCCESS,
  AUTH_PASSWORD_RESET_FAILURE,
};
