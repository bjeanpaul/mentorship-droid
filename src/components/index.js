import Text from './Text';
import Label from './Label';
import Link from './Link';
import Button from './Button';
import TextInput from './TextInput';
import MultiLineTextInput from './MultiLineTextInput';
import BaseView from './BaseView';
import Toolbar from './Toolbar';
import Header from './Header';
import Title from './Title';
import Icon from './Icon';
import HeaderIcon from './HeaderIcon';
import OverlayCompleted from './Overlay/Completed';
import OverlayLoading from './Overlay/Loading';
import Section from './Section';

export {
  BaseView,
  Toolbar,
  Text,
  Label,
  Link,
  Button,
  TextInput,
  MultiLineTextInput,
  Toolbar,
  Header,
  Title,
  Icon,
  HeaderIcon,
  OverlayCompleted,
  OverlayLoading,
  Section,
};
