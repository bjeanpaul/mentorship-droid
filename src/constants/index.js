export function generateConstants(prefix) {
  return [
    `${prefix}/request`,
    `${prefix}/success`,
    `${prefix}/failure`,
  ];
}
