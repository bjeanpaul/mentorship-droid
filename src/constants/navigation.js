// TODO rename these to not include request
const NAVIGATE_BACK_REQUEST = 'NAVIGATE_BACK_REQUEST';
const NAVIGATE_FORWARD_REQUEST = 'NAVIGATE_FORWARD_REQUEST';

const SCREEN_DISMISS = 'SCREEN_DISMISS';


export {
  NAVIGATE_BACK_REQUEST,
  NAVIGATE_FORWARD_REQUEST,
  SCREEN_DISMISS,
};
