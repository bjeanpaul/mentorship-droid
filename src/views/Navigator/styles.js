import { StyleSheet } from 'react-native';


export default StyleSheet.create({
  default: {
    flex: 1,
  },
});
