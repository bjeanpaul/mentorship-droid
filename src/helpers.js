import { find, isUndefined, partialRight as partial } from 'lodash';
import { getBaseURL } from 'src/configuration';
import { omitBy, isNull } from 'lodash';


export const omitNulls = d => omitBy(d, isNull);


export const trap = (type, fn) => v => v instanceof type
  ? Promise.resolve(v).then(fn)
  : Promise.reject(v);


export const switchError = cases => (e, ...xargs) => {
  const [type, fn] = find(cases, ([type, fn]) => e instanceof type) || [];
  return !isUndefined(fn)
    ? Promise.resolve().then(() => fn(e, ...xargs))
    : Promise.reject(e);
};


export const apiAction = ({
  method,
  request,
  success,
  failures,
}) => (...args) => (dispatch, { auth }) => Promise.resolve()
  .then(() => request(...args))
  .then(dispatch)
  .then(() => method(...args, auth))
  .then(partial(success, ...args), partial(switchError(failures), ...args))
  .then(dispatch);


export const staticAction = type => () => ({
  type
});


export const dataAction = type => data => ({
  type,
  payload: data
});


export const generateActionTypes = function generateActionNames(
  resource,
  operations = ['FETCH', 'CREATE', 'UPDATE', 'DELETE']
) {
  const actionTypes = {};
  const resourceUC = resource.toUpperCase();

  operations.forEach((operation) => {
    const operationUC = operation.toUpperCase();
    const request = `${resourceUC}_${operationUC}_REQUEST`;
    const success = `${resourceUC}_${operationUC}_SUCCESS`;
    const failure = `${resourceUC}_${operationUC}_FAILURE`;

    const operationLC = operation.toLowerCase();
    actionTypes[`${operationLC}Request`] = request;
    actionTypes[`${operationLC}Success`] = success;
    actionTypes[`${operationLC}Failure`] = failure;
  });
  return actionTypes;
};


export const filterActionTypes = function filterActionTypes(
  actionTypes,
  operation
) {
  const [request, success, failure] = [
    `${operation}Request`,
    `${operation}Success`,
    `${operation}Failure`,
  ];

  if (actionTypes[request] || actionTypes[success] || actionTypes[failure]) {
    return [
      actionTypes[request],
      actionTypes[success],
      actionTypes[failure],
    ];
  }
  return [];
};


// TODO: Delete
export const generateActionCreators = function generateActionCreators({
  resourcePath,
  actionTypes,
  requestOpts = {},
  normalizeJSON,
}) {
  const actionCreators = {};
  let url = `${getBaseURL()}/`;
  if (resourcePath) {
    url += `${resourcePath}/`;
  }

  const fetchActionTypes = filterActionTypes(actionTypes, 'fetch');
  if (fetchActionTypes.length === 3) {
    actionCreators.fetch = (onSuccess, onFailure) => ({
      type: '--generated fetch--',
      types: fetchActionTypes,
      url,
      requestOpts,
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  const createActionTypes = filterActionTypes(actionTypes, 'create');
  if (createActionTypes.length === 3) {
    actionCreators.create = ({
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated create--',
      types: createActionTypes,
      url,
      requestOpts: { ...requestOpts,
        method: 'POST',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }

  // id is required as an argument to update a record;
  // we have a scenario with profiless, where it's
  // profile/{id}/image,
  // which is awkward...
  const updateActionTypes = filterActionTypes(actionTypes, 'update');
  if (updateActionTypes.length === 3) {
    actionCreators.update = ({
      id,
      body,
      onSuccess,
      onFailure,
    }) => ({
      type: '--generated update--',
      types: updateActionTypes,
      url: `${url}${id}/`,
      requestOpts: { ...requestOpts,
        method: 'PUT',
        body: JSON.stringify(body),
      },
      normalizeJSON,
      onSuccess,
      onFailure,
    });
  }


  return actionCreators;
};
