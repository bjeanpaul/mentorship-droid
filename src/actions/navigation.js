import { staticAction } from 'src/actionHelpers';
import * as constants from 'src/constants/navigation';
import BackAndroid from 'BackAndroid';

import * as routes from 'src/constants/routes';
import { getActiveTopRoute } from 'src/store/helpers';


export const dismissScreen = staticAction(constants.SCREEN_DISMISS);


export const changeNavTab = tab => ({
  type: constants.NAV_TAB_CHANGE,
  payload: { tab },
});


export const dismissNative = () => (dispatch, ctx, getState) => {
  const state = getState();
  const activeRoute = getActiveTopRoute(state);

  switch (activeRoute.key) {
    case routes.ROUTE_LANDING:
    case routes.ROUTE_NAVIGATOR:
      return BackAndroid.exitApp();

    default:
      return dispatch(dismissScreen());
  }
};
