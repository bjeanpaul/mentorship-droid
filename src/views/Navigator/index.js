import React, { PropTypes } from 'react';
import { View } from 'react-native';

import styles from './styles';
import { NavTabBar } from 'src/components';
import { NAV_TAB_ROUTES } from 'src/constants/routes';
import * as constants from 'src/constants/navigation';


const Navigator = ({
  routes,
  activeTab,
  onTabPress,
}) => {
  const Route = routes[NAV_TAB_ROUTES[activeTab]];

  return (
    <View style={styles.default}>
      {Route && <Route />}

      <NavTabBar
        activeTab={activeTab}
        onTabPress={onTabPress}
      />
    </View>
  );
};


Navigator.propTypes = {
  routes: PropTypes.object.isRequired,
  onTabPress: PropTypes.func.isRequired,
  activeTab: PropTypes.oneOf([
    constants.NAV_TAB_ACTIVITIES,
    constants.NAV_TAB_CHAT,
    constants.NAV_TAB_COMMUNITY,
    constants.NAV_TAB_JOURNEY,
    constants.NAV_TAB_SCHEDULED_CALLS,
  ]),
};


export default Navigator;
