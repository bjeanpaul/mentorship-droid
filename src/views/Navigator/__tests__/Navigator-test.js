import { fromPairs, noop } from 'lodash';
import React from 'react';

import Navigator from 'src/views/Navigator';
import { Text } from 'src/components';
import { uidEquals } from 'app/scripts/helpers';
import { createStack, createRoute } from 'src/navigationHelpers';
import { NAV_TAB_JOURNEY, NAV_TAB_ACTIVITIES } from 'src/constants/navigation';


describe('Navigator', () => {
  const A = () => <Text>A</Text>;
  const B = () => <Text>B</Text>;

  const createComponent = (props = {}) => (
    <Navigator
      routes={{
        A,
        B,
      }}
      activeTab={NAV_TAB_ACTIVITIES}
      onTabPress={noop}
      navigationStates={fromPairs([
        [NAV_TAB_ACTIVITIES, createStack([createRoute('A')])],
        [NAV_TAB_JOURNEY, createStack([createRoute('B')])],
      ])}
      {...props}
    />);

  it('should render the currently active tab', () => {
    expect(render(createComponent({ initialActiveTab: NAV_TAB_JOURNEY })))
      .toMatchSnapshot();

    expect(render(createComponent({ initialActiveTab: NAV_TAB_ACTIVITIES })))
      .toMatchSnapshot();
  });

  it('should call onTabChange when a tab is pressed', () => {
    const el = shallow(createComponent());
    const onTabChange = jest.fn();
    el.instance().onTabChange = onTabChange;

    el.find('NavTabBar')
      .shallow()
      .findWhere(uidEquals(NAV_TAB_ACTIVITIES))
      .shallow()
      .simulate('press');

    expect(onTabChange.mock.calls).toEqual([
      [NAV_TAB_ACTIVITIES],
    ]);

    el.setState({ activeTab: NAV_TAB_ACTIVITIES })
      .find('NavTabBar')
      .shallow()
      .findWhere(uidEquals(NAV_TAB_JOURNEY))
      .shallow()
      .simulate('press');

    expect(onTabChange.mock.calls).toEqual([
      [NAV_TAB_ACTIVITIES],
      [NAV_TAB_JOURNEY],
    ]);
  });
});
