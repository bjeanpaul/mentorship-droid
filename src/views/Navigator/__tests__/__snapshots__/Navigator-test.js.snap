exports[`Navigator should render the currently active tab 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#003035",
          "fontFamily": "Brandon Text",
          "fontSize": 13,
          "textAlign": "center",
        },
        undefined,
      ]
    }>
    B
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderTopColor": "#dfe5e6",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "space-between",
      }
    }>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
</View>
`;

exports[`Navigator should render the currently active tab 2`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#003035",
          "fontFamily": "Brandon Text",
          "fontSize": 13,
          "textAlign": "center",
        },
        undefined,
      ]
    }>
    A
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderTopColor": "#dfe5e6",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "space-between",
      }
    }>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }>
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
</View>
`;
